File: lista_2/codigo_lista_2.py
#%%
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
import matplotlib.pyplot as plt
from statsmodels.tsa.arima.model import ARIMA
import statsmodels.api as sm
import seaborn as sns
import pandas as pd

#%% Leitura da base de dados
df = pd.read_excel('../datasets/treated_ger_energ_ele_hidr.xlsx')
serie_geracao = pd.Series(data=df["geracao_gwh"])
serie_geracao.index = df['mes_ano'].values

#%% Separa as bases de treinamento e teste
serie_geracao_treino = serie_geracao[:-12]
serie_geracao_teste = serie_geracao[-12:]

#%% Obtém a primeira e segunda diferença da série original
diff_1 = serie_geracao_treino.diff().dropna()
diff_2 = diff_1.diff().dropna()

# %%  FAC E FACP DAS DIFERENÇAS
def acf_pacf_grid(ts, ts_name, num_lags=50):
    """"""
    # Create a figure and axes with a 1x2 grid
    fig, axs = plt.subplots(1, 2, figsize=(10, 4))
    
    # Calculate the ACF and plot it on the left side
    acf_fig = plot_acf(ts, lags=num_lags, ax=axs[0])
    axs[0].set_title(None)

    # Calculate the PACF and plot it on the right side
    pacf_fig = plot_pacf(ts, lags=num_lags, ax=axs[1], method='ols')
    axs[1].set_title(None)

    # Set titles for each chart
    axs[0].set_xlabel('Lag')
    axs[0].set_ylabel('FAC')
    axs[1].set_xlabel('Lag')
    axs[1].set_ylabel('FACP')

    # Set a title for the grid
    fig.suptitle(f'FAC / FACP - {ts_name}')

    # Adjust the spacing between the subplots
    plt.tight_layout()

    # Show the plot
    plt.show()

# %%
def plot_chart(x, y, x_label, y_label, title, line_color='blue'):
    # Create a line plot using Seaborn
    sns.set(style="whitegrid")  # Set the style of the plot
    plt.figure(figsize=(8, 5))  # Set the size of the plot

    sns.lineplot(x=x, y=y, color=line_color)  # Use the specified line color

    # Add labels and title
    plt.xlabel(x_label)
    plt.ylabel(y_label)
    plt.title(title)

    # Add dashed red vertical lines at multiples of 12 on the x-axis
    # for i in range(len(x)):
    #     if (i + 1) % 12 == 0:
    #         plt.axvline(x=x.iloc[i], color='red', linestyle='--', linewidth=1)
    plt.show()

#%% Plot da série original - período de treino
plot_chart(
    x=serie_geracao_treino.index, 
    y=serie_geracao_treino,
    title='Geração de energia elétrica hidráulica',
    x_label='Tempo (mensal)',
    y_label='Geração (GWh)',
    line_color='xkcd:sky blue'
)

#%% Plot da 1ª primeira diferença da série temporal - período de treino
plot_chart(
    x=diff_1.index, 
    y=diff_1,
    title='1ª Diferença da série temporal',
    x_label='Tempo (mensal)',
    y_label='Geração (GWh)',
    line_color='xkcd:goldenrod'
)

#%% Plot da 2ª primeira diferença da série temporal - período de treino
plot_chart(
    x=diff_2.index, 
    y=diff_2,
    title='2ª Diferença da série temporal',
    x_label='Tempo (mensal)',
    y_label='Geração (GWh)',
    line_color='xkcd:light purple'
)

#%% Plot da FAC e FACP da segunda diferença da série
acf_pacf_grid(diff_2, '2ª Diferença da série temporal')

# %% ARMA (P,Q)
arma_models_aicc = []
# df_arma = pd.DataFrame(columns=['P Q','aicc'])

for p in range(0,6):
    for q in range (0, 6):
        print("modelo atual: {},{}".format(p,q))
        # # Versão Joao
        model = ARIMA(diff_2.values, order=(p, 0, q))
        results = model.fit()
        model_aicc = {'(p,q)': f'{p},{q}', 'AICc': results.aicc}
        arma_models_aicc.append(model_aicc)


#%% Descobre qual dos modelos ARMA(p,q) apresenta o menor valor de AICc
df_arma = pd.DataFrame(arma_models_aicc)
best_arma_aicc = df_arma.loc[df_arma['AICc'] == df_arma['AICc'].min()]['(p,q)'].iloc[0]
print('Modelo ARMA com menor AICc: ({})'.format(best_arma_aicc))
# df_arma.to_excel('ARMAPQ.xlsx', index=False) 
#%%

File: lista_2/codigo_lista_2.R
library(forecast)
library(readxl)
library(here)
library(xts)
library(openxlsx)

library(stats)
library(ggplot2)
library(forecast)
library(dplyr)

# =========================================================================

# Function to plot ACF and PACF
acf_pacf_grid <- function(ts, ts_name, num_lags = 50) {
  # Calculate ACF and PACF
  acf_result <- Acf(ts, lag.max = num_lags, plot = FALSE)
  pacf_result <- Pacf(ts, lag.max = num_lags, plot = FALSE)
  
  # Create an empty 1x2 grid
  par(mfrow = c(1, 2))

  # Plot ACF
  plot(acf_result$lag, acf_result$acf, type = 'h', main = NULL, xlab = 'Lag', ylab = 'FAC', col = 'blue')
  
  # Plot PACF
  plot(pacf_result$lag, pacf_result$acf, type = 'h', main = NULL, xlab = 'Lag', ylab = 'FACP', col = 'blue')

  # Plot a title in the center of the grid
  title(main = paste('FAC / FACP -', ts_name), outer = TRUE, line = -2, cex.main = 1.2)
}


# Function to plot a line chart
plot_chart <- function(x, y, x_label, y_label, title, line_color = 'blue', line_size = 1.5, smoothness = 0.2) {
  # Create a line plot using ggplot2
  df <- data.frame(x = x, y = y)
  ggplot(df, aes(x = x, y = y)) +
    geom_line(color = line_color, size = line_size, alpha = smoothness) +
    labs(x = x_label, y = y_label, title = title) +
    theme_minimal() +
    theme(plot.title = element_text(hjust = 0.5))
}

# =========================================================================


# Define the Excel file name
file_name <- "treated_ger_energ_ele_hidr.xlsx"

# Define the relative path to the Excel file in the "data sets" folder
file_path <- "C:/Users/João Martinez/Documents/12º Período/ENG1469 - Análise de Séries Temporais/listas_praticas/TimeSeriesAnalysis/datasets/%s" 

# Read the Excel file into a Data Frame
df <- read_excel(sprintf(file_path, file_name))

# Number of observations to reserve for the test set (last 12)
test_size <- 12

# Create the training set (all rows except the last 12)
train_set <- df[1:(nrow(df) - test_size), ]

# Create the test set (last 12 rows)
test_set <- df[(nrow(df) - test_size + 1):nrow(df), ]

# Create a time series (ts) object using xts
ts <- xts(train_set$geracao_gwh, order.by = train_set$mes_ano)

# Calculate the first difference and store it in ts_diff_1
ts_diff_1 <- diff(ts)

# Calculate the second difference and store it in ts_diff_2
ts_diff_2 <- diff(ts_diff_1)

# --------------------------------------------------------------------

# Plot the original time series
plot_chart(
  x=index(ts), y=ts,
  title='Geração de energia elétrica hidráulica',
  x_label='Tempo (mensal)',
  y_label='Geração (GWh)',
  line_size = 0.5,
  smoothness = 1.1,
  line_color='#375fa1'
)

# Plot the 1ª difference of the original time series
plot_chart(
  x=index(ts_diff_1), y=ts_diff_1,
  title='1ª Diferença da série temporal',
  x_label='Tempo (mensal)',
  y_label='Geração (GWh)',
  line_color='#32ad7a'
)

# Plot the 2ª difference of the original time series
plot_chart(
  x=index(ts_diff_2), y=ts_diff_2,
  title='2ª Diferença da série temporal',
  x_label='Tempo (mensal)',
  y_label='Geração (GWh)',
  line_color='#c26f2b'
)

# --------------------------------------------------------------------

# Plot the ACF and PACF of the 2ª diff. of the original series
acf_pacf_grid(ts_diff_2, '2ª Diferença da série temporal')

# --------------------------------------------------------------------
# Initialize lists to store ARMA models and AICc values separately
arma_list <- list()
aic_list <- list()

# Loop through different combinations of p and q
for (p in 0:5) {
  for (q in 0:5) {
    tryCatch({
      # Fit an ARMA(p, q, d) model using Arima
      model <- Arima(ts_diff_2, order=c(p, 0, q), seasonal=c(0, 0, 0, 0))
      
      # Construct the ARMA(p,q) string and append it to the arma_list
      arma_list <- append(arma_list, paste0("(", p, ",", q, ")"))
      
      # Append the AICc value to the aic_list
      aic_list <- append(aic_list, AIC(model, k = 2))  # k = 2 for AICc
    }, error = function(e) {
      # Handle any errors during model fitting
      cat("Error:", e$message, "\n")
    })
  }
}


# Create a dataframe from the two lists
results_df <- data.frame(ARMA = unlist(arma_list), AICc = unlist(aic_list))

# Print the final dataframe
print(results_df)

# Find the row index with the minimum AICc value
min_aicc_index <- which.min(results_df$AICc)

# Extract the corresponding "ARMA" value
best_arma <- results_df$ARMA[min_aicc_index]

# Print the best ARMA model order
cat("Best ARMA Model (Minimum AICc):", best_arma, "\n")


# Define the file name for the Excel file (replace with your desired file path)
excel_file <- "arma_results.xlsx"

# Write the dataframe to an Excel file
# write.xlsx(results_df, file = excel_file, sheetName = "ARMA_Results", rowNames = FALSE)

# Print a message to confirm the export
# cat("Dataframe exported to", excel_file, "\n")

    





